Posts/2021/09/Understanding Get-ACL and AD Drive Output.md

```diff
@@ -0,0 +1,245 @@
+
+    ActiveDirectoryRights : GenericAll
+    InheritanceType       : Descendents
+    ObjectType            : 00000000-0000-0000-0000-000000000000
+    InheritedObjectType   : 00000000-0000-0000-0000-000000000000
+    ObjectFlags           : None
+    AccessControlType     : Allow
+    IdentityReference     : Contoso\\User1
+    IsInherited           : False
+    InheritanceFlags      : ContainerInherit
+    PropagationFlags      : InheritOnly
+```
+
+User1 can perform an action on **OU2** and **OU1** child items but not on the **OU1** object.
+
+- **Children**: set permissions on the direct child only, not the on the object itself neither any descendants object of its children.
+You get this value when the **Applies to** set to any value other than **This object only** or **This object and all descendant objects** and at the same time the **Only apply this permission to objects and/or containers within this container** check box is selected.
+- **SelfAndChildren**: Set permissions on the object itself where the permission is placed and the direct child only.
+You get this value when the **Applies to** set to value **This object and all descendant objects** and at the same time the **Only apply this permission to objects and/or containers within this container** check box is selected.
+
+### Understanding the ObjectType Property
+
+The **ObjectType** is represented by a GUID value, even though this is one of the most important values that should be clear.
+The output of `Get-ACL` makes it complex to understand.
+The **ObjectType** is the object attribute. For example
+
+In the following output, **User1** is **Allowed** to **WriteProperty** to object **28630ebf-41d5-11d1-a9c1-0000f80367c1**.
+
+```powershell-console
+PS> (Get-Acl -Path "AD:OU=MyOrgOU,DC=Contoso,DC=Com").Access | where-Object {$_.IdentityReference -Like "Contoso\\user1"}
+
+ActiveDirectoryRights : WriteProperty
+InheritanceType       : Descendents
+ObjectType            : 28630ebf-41d5-11d1-a9c1-0000f80367c1
+InheritedObjectType   : bf967aba-0de6-11d0-a285-00aa003049e2
+ObjectFlags           : ObjectAceTypePresent, InheritedObjectAceTypePresent
+AccessControlType     : Allow
+IdentityReference     : Contoso\\User1
+IsInherited           : False
+InheritanceFlags      : ContainerInherit
+PropagationFlags      : InheritOnly
+```
+
+To resolve this GUID to a name we need to see where these GUID are stored.
+
+To get the list of *ObjectType* names, run the following PowerShell code
+
+```powershell
+$ObjectTypeGUID = @{}
+$SchGUID=Get-ADObject -SearchBase (Get-ADRootDSE).SchemaNamingContext -LDAPFilter '(SchemaIDGUID=*)' -Properties Name, SchemaIDGUID
+    Foreach ($SchemaItem in $SchGUID){
+    $ObjectTypeGUID.Add([GUID]$SchemaItem.SchemaIDGUID,$SchemaItem.Name)
+}
+
+$SchExtGUID=Get-ADObject -SearchBase "CN=Extended-Rights,$((Get-ADRootDSE).ConfigurationNamingContext)" -LDAPFilter '(ObjectClass=ControlAccessRight)' -Properties Name, RightsGUID
```

Comment:
simplify this line - create a hash table and spat the parameters to Get-ADObject